src/emulator/functionsEmulator.ts

```diff
@@ -229,7 +231,16 @@ export class FunctionsEmulator implements EmulatorInstance {
       const projectId = req.params.project_id;
 
       const reqBody = (req as RequestWithRawBody).rawBody;
-      const proto = JSON.parse(reqBody.toString());
+      let proto = JSON.parse(reqBody.toString());
+
+      if (req.headers["content-type"]?.includes("cloudevent")) {
+        // Convert request payload to CloudEvent.
+        // TODO(taeold): Converting request payload to CloudEvent object should be done by the functions runtime.
+        // However, the Functions Emulator communicates with the runtime via socket not HTTP, and CE metadata
+        // embedded in HTTP header may get lost. Once the Functions Emulator is refactored to communicate to the
+        // runtime instances via HTTP, move this logic there.
+        proto = HTTP.toEvent({ headers: req.headers, body: proto });
```

Comment:
I have mixed feelings about the CloudEvents SDK (partially just because I was strongly against the committee including canonical SDKs). If we're going to put it in the prod SDK we should see how big the SDK is. AFAIK functions framework doesn't use the CloudEvents library because we always know the encoding that our clients will use.

Also, I just take issue that `toEvent` calls `Object.freeze` IIRC, which means that if we were to use this in `firebase-functions` we'd have to clone the object just so we could massage `event.data` from a raw wire protocol into an SDK type.

Reply:
Thanks for flagging this. Answering couple of issues you raised here for the record:

1. cloudevents npm package is [pretty big](https://bundlephobia.com/package/cloudevents@4.0.3) at 145.9KB minified. In contrast, firebase-functions SDK is is [115.7KB](https://bundlephobia.com/package/firebase-functions@3.15.7).

2. Functions Framework seems to take a clever shortcut for converting binary encoded cloudevent to structured one. Going to copy their methods https://github.com/GoogleCloudPlatform/functions-framework-nodejs/blob/9cf46edc9b1b890e2a6fd1aec8d818074585783b/src/invoker.ts#L146-L149.